File: frontend/src/api/teams.ts
import type { User } from 'firebase/auth';
import { fetchWithAuth, readErrorDetail } from './http';
import type { UserTeamSummary } from './workspace';

export type TeamInviteStatus = 'active' | 'expired' | 'revoked';

export type TeamInvite = {
  id: string;
  team_id: string;
  created_by: {
    id: string;
    name: string;
  };
  created_at: string;
  expires_at: string;
  status: TeamInviteStatus;
  can_revoke: boolean;
};

export type InvitePreview = {
  team_id: string;
  team_name: string;
  expires_at: string;
  already_member: boolean;
};

export async function createTeam(user: User, name: string): Promise<UserTeamSummary> {
  const response = await fetchWithAuth(user, '/api/teams', {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
  if (!response.ok) throw new Error(await teamError(response));
  const payload = (await response.json()) as { team?: ApiTeam };
  if (!payload.team) throw new Error('チーム作成APIの応答を読み取れませんでした。');
  return toTeam(payload.team);
}

export async function createDemoTeam(user: User): Promise<UserTeamSummary> {
  const response = await fetchWithAuth(user, '/api/teams/demo', { method: 'POST' });
  if (!response.ok) throw new Error(await teamError(response));
  const payload = (await response.json()) as { team?: ApiTeam };
  if (!payload.team) throw new Error('デモモードAPIの応答を読み取れませんでした。');
  return toTeam(payload.team);
}

export async function fetchTeamInvites(user: User, teamId: string): Promise<TeamInvite[]> {
  const response = await fetchWithAuth(user, `/api/teams/${teamId}/invites`);
  if (!response.ok) throw new Error(await teamError(response));
  const payload = (await response.json()) as { invites?: TeamInvite[] };
  if (!Array.isArray(payload.invites)) throw new Error('招待一覧APIの応答を読み取れませんでした。');
  return payload.invites;
}

export async function createTeamInvite(
  user: User,
  teamId: string,
): Promise<{ invite: TeamInvite; inviteUrl: string }> {
  const response = await fetchWithAuth(user, `/api/teams/${teamId}/invites`, { method: 'POST' });
  if (!response.ok) throw new Error(await teamError(response));
  const payload = (await response.json()) as { invite?: TeamInvite; invite_url?: string };
  if (!payload.invite || !payload.invite_url) throw new Error('招待作成APIの応答を読み取れませんでした。');
  return { invite: payload.invite, inviteUrl: payload.invite_url };
}

export async function revokeTeamInvite(user: User, teamId: string, inviteId: string): Promise<void> {
  const response = await fetchWithAuth(user, `/api/teams/${teamId}/invites/${inviteId}`, { method: 'DELETE' });
  if (!response.ok) throw new Error(await teamError(response));
}

export async function fetchInvitePreview(user: User, token: string): Promise<InvitePreview> {
  const response = await fetchWithAuth(user, `/api/invites/${encodeURIComponent(token)}`);
  if (!response.ok) throw new Error(await teamError(response));
  return (await response.json()) as InvitePreview;
}

export async function acceptTeamInvite(user: User, token: string): Promise<UserTeamSummary> {
  const response = await fetchWithAuth(user, `/api/invites/${encodeURIComponent(token)}/accept`, { method: 'POST' });
  if (!response.ok) throw new Error(await teamError(response));
  const payload = (await response.json()) as { team?: ApiTeam };
  if (!payload.team) throw new Error('招待参加APIの応答を読み取れませんでした。');
  return toTeam(payload.team);
}

type ApiTeam = {
  id: string;
  name: string;
  role: UserTeamSummary['role'];
  member_count: number;
};

function toTeam(team: ApiTeam): UserTeamSummary {
  return {
    team_id: team.id,
    name: team.name,
    role: team.role,
    member_count: team.member_count,
  };
}

async function teamError(response: Response): Promise<string> {
  const detail = await readErrorDetail(response);
  if (detail === 'team name is required') return 'チーム名を入力してください。';
  if (detail === 'team name is too long') return 'チーム名は255文字以内で入力してください。';
  if (detail === 'invite is expired') return 'この招待リンクの有効期限は切れています。';
  if (detail === 'invite is revoked') return 'この招待リンクは無効化されています。';
  if (detail === 'invite not found') return '招待リンクが見つかりません。';
  if (detail === 'host email is required') return 'ログインユーザーのメールアドレスを確認できませんでした。';
  if (detail === 'aiboard api is not configured') return 'ミーティング作成機能が設定されていません。';
  if (detail === 'aiboard api key was rejected') return 'ミーティング作成機能の認証に失敗しました。';
  if (detail.startsWith('aiboard request validation failed')) return 'デモミーティングを作成できませんでした。';
  if (detail === 'failed to create aiboard meeting' || detail === 'invalid aiboard response') {
    return 'デモミーティングを作成できませんでした。';
  }
  if (response.status === 403) return 'この操作を行う権限がありません。';
  return '処理に失敗しました。時間をおいて再試行してください。';
}

File: backend/app/api/teams.py
from datetime import datetime
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Response, status
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.auth import AuthenticatedUser, verify_firebase_user
from app.core.config import get_settings
from app.db.session import get_db_session
from app.services.aiboard_service import (
    AiboardAuthenticationError,
    AiboardConfigurationError,
    AiboardRequestError,
)
from app.services.demo_mode_service import create_demo_mode_workspace
from app.services.team_invite_service import (
    InviteSummary,
    TeamSummary,
    accept_team_invite,
    create_team,
    create_team_invite,
    delete_team,
    get_invite_preview,
    list_team_invites,
    revoke_team_invite,
)

router = APIRouter()
settings = get_settings()


class TeamResponse(BaseModel):
    id: UUID
    name: str
    role: str
    member_count: int


class CreateTeamRequest(BaseModel):
    name: str = Field(max_length=255)


class CreateTeamResponse(BaseModel):
    team: TeamResponse


class DeleteTeamRequest(BaseModel):
    name: str = Field(max_length=255)


class InviteCreatorResponse(BaseModel):
    id: UUID
    name: str


class InviteResponse(BaseModel):
    id: UUID
    team_id: UUID
    created_by: InviteCreatorResponse
    created_at: datetime
    expires_at: datetime
    status: str
    can_revoke: bool


class InviteListResponse(BaseModel):
    invites: list[InviteResponse]


class CreateInviteResponse(BaseModel):
    invite: InviteResponse
    invite_url: str


class InvitePreviewResponse(BaseModel):
    team_id: UUID
    team_name: str
    expires_at: datetime
    already_member: bool


class AcceptInviteResponse(BaseModel):
    team: TeamResponse


def to_team_response(team: TeamSummary) -> TeamResponse:
    return TeamResponse(id=team.id, name=team.name, role=team.role, member_count=team.member_count)


def to_invite_response(invite: InviteSummary) -> InviteResponse:
    return InviteResponse(
        id=invite.id,
        team_id=invite.team_id,
        created_by=InviteCreatorResponse(id=invite.created_by_user_id, name=invite.created_by_name),
        created_at=invite.created_at,
        expires_at=invite.expires_at,
        status=invite.status,
        can_revoke=invite.can_revoke,
    )


@router.post("/teams", response_model=CreateTeamResponse, status_code=status.HTTP_201_CREATED)
async def post_team(
    request: CreateTeamRequest,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> CreateTeamResponse:
    team = await create_team(session=session, auth_user=auth_user, name=request.name)
    return CreateTeamResponse(team=to_team_response(team))


@router.post("/teams/demo", response_model=CreateTeamResponse, status_code=status.HTTP_201_CREATED)
async def post_demo_team(
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> CreateTeamResponse:
    host_email = (auth_user.email or auth_user.claims.get("email") or "").strip()
    if not host_email:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="host email is required",
        )
    try:
        result = await create_demo_mode_workspace(
            session=session,
            auth_user=auth_user,
            aiboard_api_base_url=settings.aiboard_api_base_url,
            aiboard_api_key=settings.aiboard_api_key,
        )
    except AiboardConfigurationError as exc:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(exc),
        ) from exc
    except AiboardAuthenticationError as exc:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=str(exc),
        ) from exc
    except AiboardRequestError as exc:
        await session.rollback()
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=str(exc),
        ) from exc
    return CreateTeamResponse(team=to_team_response(result.team))


@router.delete("/teams/{team_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_team_endpoint(
    team_id: UUID,
    request: DeleteTeamRequest,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> Response:
    await delete_team(
        session=session,
        auth_user=auth_user,
        team_id=team_id,
        confirmation_name=request.name,
    )
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/teams/{team_id}/invites", response_model=InviteListResponse)
async def get_team_invites(
    team_id: UUID,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> InviteListResponse:
    invites = await list_team_invites(session=session, auth_user=auth_user, team_id=team_id)
    return InviteListResponse(invites=[to_invite_response(invite) for invite in invites])


@router.post("/teams/{team_id}/invites", response_model=CreateInviteResponse, status_code=status.HTTP_201_CREATED)
async def post_team_invite(
    team_id: UUID,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> CreateInviteResponse:
    invite, token = await create_team_invite(session=session, auth_user=auth_user, team_id=team_id)
    return CreateInviteResponse(
        invite=to_invite_response(invite),
        invite_url=f"{settings.frontend_base_url.rstrip('/')}/invite/{token}",
    )


@router.delete("/teams/{team_id}/invites/{invite_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_team_invite(
    team_id: UUID,
    invite_id: UUID,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> Response:
    await revoke_team_invite(session=session, auth_user=auth_user, team_id=team_id, invite_id=invite_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.get("/invites/{token}", response_model=InvitePreviewResponse)
async def get_invite(
    token: str,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> InvitePreviewResponse:
    preview = await get_invite_preview(session=session, auth_user=auth_user, token=token)
    return InvitePreviewResponse(
        team_id=preview.team_id,
        team_name=preview.team_name,
        expires_at=preview.expires_at,
        already_member=preview.already_member,
    )


@router.post("/invites/{token}/accept", response_model=AcceptInviteResponse)
async def post_invite_accept(
    token: str,
    auth_user: AuthenticatedUser = Depends(verify_firebase_user),
    session: AsyncSession = Depends(get_db_session),
) -> AcceptInviteResponse:
    team = await accept_team_invite(session=session, auth_user=auth_user, token=token)
    return AcceptInviteResponse(team=to_team_response(team))
